errors/error.go

```diff
@@ -0,0 +1,268 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package errors implements the Terramate standard error type.
+// It's heavily influenced by Rob Pike `errors` package in the Upspin project:
+// 	https://commandcenter.blogspot.com/2017/12/error-handling-in-upspin.html
+package errors
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+)
+
+// Error is the default Terramate error type.
+// It may contains some fields unset, but not all.
+// See E() for its usage.
+type Error struct {
+	// Kind is the kind of error.
+	Kind Kind
+
+	// Description of the error.
+	Description string
+
+	// FileRange holds the error source.
+	FileRange hcl.Range
+
+	// Stack which originated the error.
+	Stack Stack
```

Comment:
hmm why not just use the stack.S type ? then we could create an error related to a stack passing the stack itself, getting a string will allow different parts of the code to pass a different representation of the stack, like some code passing the name, other part passing the absolute path, etc.

Reply:
Because `errors` must be imported by every other package, then importing any other TM package here will make cycles...

We can add a `filepath.IsAbs()` check so it helps catch this.

Reply:
As discussed out of the PR, we can also go with an interface to break the cycle, which seems nice IMO, so we can still pass a stack directly when creating errors for it + no need to check things for invariants since the stack itself already did that.